### ISSUE TITLE
Annotation to when BOOL indicates to call GetLastError

### ISSUE THREAD

**Issue by rylev** [microsoft/win32metadata#536]
A common pattern is for functions to return a `BOOL` and if `FALSE`, for the caller to check the return value of `GetLastError`. It would be potentially be help for the metadata to annotate such uses such that consumers of the metadata could provide more useful return types of such functions.

For instance, in Rust instead of the following:

```rust
let result: BOOL = SomeMethod();
if !result.as_bool() {
   return Err(GetLastError());
}
```

You could have:

```rust
let result: Result<(), WIN32_ERROR> = SomeMethod();
// Users would normally just do:
SomeMethod()?;
```

**Comment by KalleOlaviNiemitalo** [comment]
NULL and INVALID_HANDLE_VALUE are other popular values to return when a function sets the last error. Does the metadata describe those yet?

Then there are also functions like [CreateMutexW](https://docs.microsoft.com/windows/win32/api/synchapi/nf-synchapi-createmutexw) and [CreateSemaphoreW](https://docs.microsoft.com/windows/win32/api/synchapi/nf-synchapi-createsemaphorew), which can return a valid handle but set ERROR_ALREADY_EXISTS as the last error, to indicate that they opened an existing object rather than created a new one.

**Comment by mikebattista** [comment]
@kennykerr @AArnott @sotteson1 

We have `[DllImport(..., SetLastError = true)]` on all of these examples. How are the projections using that information today? If a function returns `bool`, and `SetLastError = true`, is that enough of a hint for projections to provide the friendly interface? Likewise when a function has `SetLastError = true` and returns `NULL`.

The case where a function returns a non-`NULL` invalid value like `INVALID_HANDLE_VALUE` but also sets the error would probably require some kind of annotation regardless.

I'm not sure about the CreateMutex and CreateSemaphore cases.

**Comment by kennykerr** [comment]
Ah, I think `SetLastError = true` in C# equates to the `SupportsLastError(0x0040)` flag in the `ImplMap` table. I should be able to use that, assuming there's an unambiguous way to determine whether the function succeeded/failed. We can transfer this issue to the `windows-rs` repo. I'm assuming the C# projection already supports this automatically.

**Comment by mikebattista** [comment]
Given that we have a `HANDLE` type, you could probably check for `INVALID_HANDLE_VALUE` in addition to `NULL` in that case as well without any additional annotations in the metadata. [CreateFile](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea) is an example like that.

**Comment by AArnott** [comment]
`SetLastError` is critical for .NET projections, but it doesn't tell us *when* to call `GetLastError()`. Calling it in success cases can be a mistake because not all methods call `SetLastError` in success paths. For Rust (or C# if we were to generate our own throwing friendly overloads) we would need to know how to discern a successful call from a failing one. That info isn't in the header files so it would be tedious to add that to the metadata. 
Short of the metadata having the authoritative answer, I would tend to shy away from assuming common conventions are always true (like a `false` result suggesting failure) since as I mentioned above, calling `GetLastError` in non-error cases can be very misleading.

**Comment by kennykerr** [comment]
Yes, it depends on whether we can clearly tell whether a call succeeded/failed.

**Comment by mikebattista** [comment]
So for cases where `SetLastError = true` and projections can clearly identify failure (like HANDLE returning INVALID_HANDLE_VALUE), they can provide a friendlier interface.

Regarding annotations for the failure values of functions, I agree on the surface it sounds like that would be pretty tedious to capture. I'm open to add support for an attribute or something, but we'd need to rely on the community contributing to populating that across the API surface.

Is that something we should add to the backlog? Or are projections fine with the information currently captured?

**Comment by AArnott** [comment]
I have no plans in CsWin32 to add throwing-overloads, so we have everything we need. 
For COM interfaces, HRESULT is always returned and error cases are clearly spec'd, so throwing from those is already built into the CLR's interop layer (when PreserveSig is false). 

I don't object to an attribute either, even if contributions alone supply it. But we'd have to first identify the possible method outputs that may indicate an error so that the attribute can be designed to express all of them.

**Comment by MarijnS95** [comment]
Specifically linking in https://github.com/microsoft/windows-rs/issues/3538, it appears that `windows-rs` already handles `BOOL` return types with `FALSE` when they have the `SetLastError = true` attribute (that `SupportsLastError` bit).

Perhaps the issue title is updated, because this is not yet supported for other types.

As described in https://github.com/microsoft/windows-rs/issues/3538, SAL defines a very specific attribute to configure what return value - in the case of `DWORD` types for example - result in an error that can be retrieved with `GetLastError` when the matching `SupportsLastError` bit is also set in the metadata. An example of such an annotation is `_Success_(return != 0)` found at https://github.com/microsoft/windows-rs/issues/3229#issuecomment-2715785195 / https://github.com/microsoft/win32metadata/blob/a4418f7f3c58becfbb32b448d21ff82023d8bcb9/generation/WinSDK/RecompiledIdlHeaders/um/Psapi.h#L194-L205 (plus some `_When_` / `_Out_writes_xx_` to describe how many bytes are written by the returned `DWORD`).

Is this something we can support in one way or another? Having such code expressions in attributes might be tricky, but maybe we can start translating a few general cases where an error happens if the value is specifically zero or non-zero?

**Comment by riverar** [comment]
How do envision projections determine whether the function has failed or succeeded?

**Comment by MarijnS95** [comment]
@riverar the sample function above quite clearly returns a `DWORD` with `_Success_(return != 0)` annotation. Hence the windows-rs projection can simply do something like:

```rust
NonZero::<u32>::new(GetModuleFileNameExA()).ok_or(Error::from_win32)
```
?

**Comment by MarijnS95** [comment]
To be clear, I'm not suggesting us to immediately be able to handle all expressions in `_Success_()` but the base `return != 0` case should be common and easy to handle?

I do acknowledge that requires yet many more missing SAL annotation overrides in this project, the file linked above for example lacks it for all other `GetXxxNameX` functions besides the `GetModuleFileNameExX()` sample.

**Comment by MarijnS95** [comment]
I'm not envisioning anything, just requesting the team here to take care of this in their metadata 😉 

As *already suggested*, you could start with the `return != 0` base case and expand to more complicated expressions later.
